### PR TITLE
Remove transitive requirements when swapping children

### PR DESCRIPTION
**Pull request by danielsdeleo**
- [x] CHANGELOG
- [x] Tests

---

I'm investigating an issue (https://github.com/berkshelf/solve/issues/57) in the solve library, which is using Molinillo. I was able to add a reproduction case (https://github.com/berkshelf/solve/commit/b84500e80c22e0596b530838f7a7d299f6f9f870) as a RSpec test for solve, which fails with the following:

```
  1) Solutions when using the ruby solver repro case?
     Failure/Error: activated.vertex_named(name).payload ? 0 : 1,
     
     NoMethodError:
       undefined method `payload' for nil:NilClass
     # ./lib/solve/ruby_solver.rb:145:in `block in sort_dependencies'
     # ./lib/solve/ruby_solver.rb:142:in `each'
     # ./lib/solve/ruby_solver.rb:142:in `sort_by'
     # ./lib/solve/ruby_solver.rb:142:in `sort_dependencies'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:131:in `block (2 levels) in <class:Resolution>'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:431:in `push_state_for_requirements'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:423:in `require_nested_dependencies_for'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:411:in `activate_spec'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:353:in `attempt_to_swap_possibility'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:333:in `attempt_to_activate_existing_spec'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:318:in `attempt_to_activate'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:147:in `process_topmost_state'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolution.rb:72:in `resolve'
     # /Users/ddeleo/contrib/Molinillo/lib/molinillo/resolver.rb:42:in `resolve'
     # ./lib/solve/ruby_solver.rb:198:in `resolve_with_error_wrapping'
     # ./lib/solve/ruby_solver.rb:75:in `resolve'
     # ./lib/solve.rb:64:in `it!'
     # ./spec/acceptance/ruby_solver_solutions_spec.rb:300:in `block (2 levels) in <top (required)>'
```

When I enable Molinillo's debugging and add some printf debugging of my own, what I find is that a requirement is added for a grandchild, but the child later gets swapped. The `#detach_vertex_named` method is recursive, so it removes the grandchild's vertex, but the requirements removal is not recursive, so there's a "zombie" requirement, so that `activated.vertex_named(name)` returns `nil`.

In more concrete terms (these are all Chef cookbooks), there is a cookbook named "build-essential". The old version of this depends on a cookbook named "7-zip"; the new version of it depends on a replacement named "seven_zip". Both "7-zip" and "seven_zip" depend on a cookbook called "windows". When solving the dependencies, Molinillo at first picks the solution using the new version of "build-essential", and activates "seven_zip" and "windows". This conflicts with a version requirement elsewhere (we need the older "build-essential"), so Molinillo calls `#fixup_swapped_children` to remove "seven_zip". In the meantime, the requirements array has collected a requirement for the windows cookbook. Since only the vertices are recursively removed, the windows requirement stays there, until it triggers the exception case for `activated.vertex_named("windows") #=> nil`.

As for the code in this pull request:
1. I'm not sure if this is the proper approach to solving the problem; I don't have a good understanding of most of the moving parts here
2. Even if the approach is right, it's only a partial solution for the reason I wrote in the code comment.
3. In any case there of course should be a test, but my repro case is a little picky based on the order things are returned from `#sort_dependencies` and I haven't yet looked at how the JSON fixtures work.

Despite that, it does pass Molinillo's specs, and it fixes my repro case.

### CODE REVIEW

**Comment by segiddins** [comment]
Thanks! We ought to add either an integration spec or a more unit-y test to cover this, and the build will have to go green.

**Comment by danielsdeleo** [comment]
Ok, the build failure seems to just be rubocop complaining about my `TODO` comment, which I would remove anyway.

Before I go any further, I just want to be extra sure that this approach is safe (i.e., won't cause a regression for any other case). What I'm worried about is a case like:

```
A   E
|   |
|   |
B   F
 \ /
  C
  |
  D
```

Is there a case where `F` could get swapped, and therefore my change would remove the (valid) requirements on `C` and `D` ? My apologies if this is a daft question, as this is my first foray into the implementation.

File: lib/molinillo/resolution.rb [hunk 1]
```diff
@@ -364,6 +364,16 @@ def fixup_swapped_children(vertex)
...
             activated.detach_vertex_named(succ.name)
             requirements.delete_if { |r| name_for(r) == succ.name }
+            # TODO:
+            # 1. I'm not sure if it's correct to be deleting these
+            # requirements. Does not supporting circular deps ensure that the
+            # requirement must have come from `succ`? What about diamond-shaped
+            # dependency graphs?
+            # 2. If it's okay to delete the requirements, this should be
+            # recursive.
+            succ.successors.each do |succ_dep|
```
Comment (segiddins) [inline lib/molinillo/resolution.rb]: are you sure this shouldn't be `predecessors`? also, this can be folded into the prior `delete_if` loop by using an `include?` check

**Comment by segiddins** [comment]
@danielsdeleo it _should_ be fine. When I have some time I'll run this change through bundler are CocoaPods to double check though

File: lib/molinillo/resolution.rb (hunk 1 above)
Comment (danielsdeleo) [inline lib/molinillo/resolution.rb]: Definitely successors. In my repro case, I have: `build-essential(new) -> seven_zip -> windows` and `build-essential(old) -> 7-zip -> windows`. It's when Molinillo detects that it needs to replace `build-essential(new)` with `build-essential(old)` that we hit this code path; in this case, `windows` is the "zombie" requirement, it's a successor of `seven_zip`.

And yeah, per the comments above, this is a bad patch as currently written, I just wanted to get your opinion on the general approach so I didn't waste effort on doing the wrong thing. I'll re-write it.

Comment (segiddins) [inline lib/molinillo/resolution.rb]: It seems fine, but you might just want to make this 'deleting successor requirements' thing recursive ?

**Comment by danielsdeleo** [comment]
@segiddins re-worked the patch to make it correctly recurse through the successors and avoid extra loops when removing requirements.

As for a test, I'll have to spend some time learning how the integration tests work. My repro case is dependent on processing the dependencies in a specific order to ensure it hits the `fixup_swapped_children` path; in Solve we use `activated.vertex_named(name).payload ? 0 : graph.versions(dependency.name).count` in `#sort_dependencies` which we can use to control the ordering.

**Comment by segiddins** [comment]
@danielsdeleo feel free to make it a unit test instead of an integration spec if that's easier!

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -33,4 +33,6 @@
 RSpec.configure do |config|
   # Enable flags like --only-failures and --next-failure
   config.example_status_persistence_file_path = '.rspec_status'
+  config.filter_run focus: true
+  config.run_all_when_everything_filtered = true
```
Comment (danielsdeleo) [inline spec/spec_helper.rb]: Dunno if y'all don't like this for some reason, would prefer it be added in a different PR or whatever. Easy to remove with a rebase.

File: spec/swap_child_with_successors.json [hunk 3]
```diff
@@ -0,0 +1,123 @@
+{
```
Comment (danielsdeleo) [inline spec/swap_child_with_successors.json]: Obviously this is a hack, wanted to get a test in the PR for review before dealing with submodules.

File: spec/resolver_spec.rb [hunk 4]
```diff
@@ -119,6 +119,52 @@ def initialize(fixture_path)
...
         expect(resolved.map(&:payload).map(&:to_s)).to eq(['actionpack (1.2.3)'])
       end
+
+      class SwapChildWithSuccessorsIndex < TestIndex
+
+        # The bug we want to write a regression test for only occurs when
+        # Molinillo processes dependencies in a specific order for the given
+        # index and demands. This sorting logic ensures we hit the repro case
+        # when using the index file "swap_child_with_successors"
+        def sort_dependencies(dependencies, activated, conflicts)
```
Comment (danielsdeleo) [inline spec/resolver_spec.rb]: Not sure how best to handle this. The sorting algorithm in `TestIndex` changes the order and sidesteps the bug, so I copied the sorting we're using in Solve to make the repro case work.

Comment (segiddins) [inline spec/resolver_spec.rb]: this is probably fine :)

**Comment by danielsdeleo** [comment]
@segiddins ended up with sort of a hybrid test but it probably needs cleaning up in some way

File: spec/resolver_spec.rb (hunk 4 above)
Comment (segiddins) [inline spec/resolver_spec.rb]: (though I'd prefer to make the class anonymous in the spec itself)

Comment (danielsdeleo) [inline spec/resolver_spec.rb]: Okay, that's done

**Comment by danielsdeleo** [comment]
Made a PR for adding my fixtures into the other repo: https://github.com/CocoaPods/Resolver-Integration-Specs/pull/4

**Comment by danielsdeleo** [comment]
Ok, I think the test should be good now. I see you added a todo item for a changelog entry. Should I do that or do you do that?

**Comment by segiddins** [comment]
it'd be great if you could :)

**Comment by danielsdeleo** [comment]
Done.

**Comment by segiddins** [comment]
Sweeeet. Thanks so much!

@homu r+

**Comment by danielsdeleo** [comment]
Thanks! Quick question, do you plan to release this soonish? I need to figure out if I should just wait for a release or investigate integrating this into our build process (we build native OS packages and our build process makes it a bit more involved to pull dependencies from git compared to just changing the source in a Gemfile).

**Comment by segiddins** [comment]
@danielsdeleo plan on releasing this either today or tomorrow

**Comment by danielsdeleo** [comment]
Awesome, thanks!
